### ISSUE TITLE
[Bug]: RAG pipeline not updating contexts after 1st query [RAG]

### ISSUE THREAD

**Issue by swaroop11** [microsoft/autogen#2276]
### Describe the bug

I have created a .py file with below code which uses `RetrieveUserProxyAgent` and `RetrieveUserProxyAgent` agents. 
I have kept all the settings to default only and initiated the chat with the `ragproxyagent`.
For the first question it correctly pulls context and provide answer, but from next query onwards it just say "UPDATE CONTEXT" and generates no response.

Ideally, it should pull new context for every next query user is asking and generate response based on that, but not doing that.

### Steps to reproduce

```import os
import autogen
from autogen.agentchat.assistant_agent import AssistantAgent
from autogen.agentchat.contrib.retrieve_assistant_agent import RetrieveAssistantAgent
from autogen.agentchat.contrib.retrieve_user_proxy_agent import RetrieveUserProxyAgent
from langchain.text_splitter import RecursiveCharacterTextSplitter

config_list = autogen.config_list_from_json("OAI_CONFIG_LIST",
                                            filter_dict={"model": ["gpt-4"],},)

gpt4_api_key = config_list[0]["api_key"]
os.environ['OPENAI_API_KEY'] = gpt4_api_key


text_splitter = RecursiveCharacterTextSplitter(chunk_size=00, chunk_overlap=0, separators=["\n"], keep_separator=False)

assistant = AssistantAgent(
    name="assistant",
    llm_config={
        "timeout": 600,
        "config_list": config_list,
    },
)

ragproxyagent = RetrieveUserProxyAgent(
    name="ragproxyagent",
    retrieve_config={
        "task": "qa",
        "custom_text_split_function": text_splitter.split_text,
        "docs_path": "context_file.txt",
    },
    code_execution_config=False,
)

assistant.reset()
ragproxyagent.initiate_chat(assistant, message=ragproxyagent.message_generator, problem="i want to stop getting notification what to do?",
                            `n_results=2)


### Model Used

gpt-4 

### Expected Behavior

It should have called RAG with new context for each input query, but that is not happening.

### Screenshots and logs

These are the logs, I removed sensitive info.
![image](https://github.com/microsoft/autogen/assets/37260686/2f55a4e4-70a3-4ff3-a303-6c8fd47e2b15)


### Additional Information

_No response_

**Comment by thinkall** [comment]
Hi @swaroop11 ,  in the current design, if you'd like to ask for a series of questions, the best design would be:
```python
while True:
    problem = input("Please enter your question: ")
    if problem == "":
        break
    assistant.reset()
    ragproxyagent.initiate_chat(assistant, message=ragproxyagent.message_generator, problem=problem)
```

**Comment by swaroop11** [comment]
Hi @thinkall, I checked your suggestion but it giving the same issue.
If `human_input_mode="ALWAYS"` then RAG is performed only once and agent tries to answer all the future questions using that context only. Ideally it should pull new context for each input query. So the issue is not solved

`ragproxyagent = RetrieveUserProxyAgent(
    name="ragproxyagent",
    human_input_mode="ALWAYS",
    max_consecutive_auto_reply=2,
    retrieve_config={
        "task": "qa",
        "custom_text_split_function": text_splitter.split_text,
        "docs_path": "context_file.txt",
    },
    code_execution_config=False,
)
`

**Comment by Josephrp** [comment]
> If `human_input_mode="ALWAYS"` then RAG is performed only once and agent tries to answer all the future questions using that context only. Ideally it should pull new context for each input query. So the issue is not solved

hey there @swaroop11 autogen is a bit wierd because it didnt start with rag, and i do think it "inherited" some idiosyncracies there, for example the importance of a good system prompt.

check out this demo on huggingface to see a good example for that. hope this helps : https://huggingface.co/spaces/thinkall/autogen-demos

**Comment by thinkall** [comment]
> Hi @thinkall, I checked your suggestion but it giving the same issue. If `human_input_mode="ALWAYS"` then RAG is performed only once and agent tries to answer all the future questions using that context only. Ideally it should pull new context for each input query. So the issue is not solved
> 
> `ragproxyagent = RetrieveUserProxyAgent( name="ragproxyagent", human_input_mode="ALWAYS", max_consecutive_auto_reply=2, retrieve_config={ "task": "qa", "custom_text_split_function": text_splitter.split_text, "docs_path": "context_file.txt", }, code_execution_config=False, ) `

Hi @swaroop11 , When you set human_input_mode="ALWAYS", for each question, when assistant replied, you will be prompted to input your reply (human_input); after the interaction finished, you'll be prompted to input the next question and start a new round of chat.

**Comment by julie2025** [comment]
Hi, any update on this issue? or is it that the context update does not work if human_input_mode is set to ALWAYS? is there any workaround for this? thanks in advance.

**Comment by thinkall** [comment]
Hi @swaroop11 , in  the example notebook `notebook/agentchat_RetrieveChat.ipynb`, I tried running below code,
```python
while True:
    problem = input("Please enter your question: ")
    if problem == "":
        break
    assistant.reset()
    ragproxyagent.initiate_chat(assistant, message=ragproxyagent.message_generator, problem=problem)
```
The outputs:
```
2024-07-22 18:04:40,295 - autogen.agentchat.contrib.retrieve_user_proxy_agent - INFO - Use the existing collection `autogen-docs`.
Trying to create collection.
2024-07-22 18:04:40,854 - autogen.agentchat.contrib.retrieve_user_proxy_agent - INFO - Found 2 chunks.
Number of requested results 20 is greater than number of elements in index 2, updating n_results = 2
Model gpt4-1106-preview not found. Using cl100k_base encoding.
VectorDB returns doc_ids:  [['bdfbc921', '7968cf3c']]
Adding content of doc bdfbc921 to context.
Model gpt4-1106-preview not found. Using cl100k_base encoding.
Adding content of doc 7968cf3c to context.

...

Number of requested results 20 is greater than number of elements in index 2, updating n_results = 2
Model gpt4-1106-preview not found. Using cl100k_base encoding.
VectorDB returns doc_ids:  [['7968cf3c', 'bdfbc921']]
Adding content of doc 7968cf3c to context.
Model gpt4-1106-preview not found. Using cl100k_base encoding.
Adding content of doc bdfbc921 to context.
...
```


The results clearly showed that for two questions `How can I use FLAML to perform a classification task and use spark to do parallel training. Train 30 seconds and force cancel jobs if time limit is reached.` and `Who is the author of FLAML?`, the order of the retrieved docs are different. If you have only a few docs, since the context lenght is very short, it can happen that all docs are inserted into the context, but the retrieve process will run for each question.

@julie2025 , do you see the same issue? Based on the current info in this conversation,  I think it can be addressed w/o updating AutoGen code.

**Comment by julie2025** [comment]
Hi @thinkall, Thank you for answering. Indeed, this works, but this makes it difficult to handle the context/conversation history and the token count/token cost handling, which are already implemented when using the human in the loop option. I think the proposal works as a workaround, but it would be better if this is fixed in Autogen directly.

**Comment by thinkall** [comment]
> Hi @thinkall, Thank you for answering. Indeed, this works, but this makes it difficult to handle the context/conversation history and the token count/token cost handling, which are already implemented when using the human in the loop option. I think the proposal works as a workaround, but it would be better if this is fixed in Autogen directly.

Thank you @julie2025 for your feedback. It's easy to handle token count/cost, you can try `chat_result=ragproxyagent.initiate_chat(assistant, message=ragproxyagent.message_generator, problem=problem)` and `chat_result.cost` will include below info:
```
{'usage_including_cached_inference': {'total_cost': 0.06347,
  'gpt-4o-2024-05-13': {'cost': 0.06347,
   'prompt_tokens': 11323,
   'completion_tokens': 457,
   'total_tokens': 11780}},
 'usage_excluding_cached_inference': {'total_cost': 0.06347,
  'gpt-4o-2024-05-13': {'cost': 0.06347,
   'prompt_tokens': 11323,
   'completion_tokens': 457,
   'total_tokens': 11780}}}
```

Do you want to support different chat topics in one chat session? I'm not sure if it's good to mix different topics in one session, not only unreleated histories will distract LLMs to answer a specific question, but also will increase the token usage.

**Comment by thinkall** [comment]
Close as no active response for a long time. Please reopen it or create a new issue if needed.
